Display default permissions of all modules even if they are deactivat…

yurabakhtin opened the pull request:
…ed by default

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [x] Changelog was modified

**Other information:**
Issue: https://github.com/humhub/humhub/issues/4828

luke- reviewed (changes requested):
I get following error, since the method is not set to `public`.
![image](https://user-images.githubusercontent.com/4736168/106917045-269aa380-6708-11eb-9396-46ec81ddc04c.png)

What do you think about this approach? here we would not need to change any modules:
https://github.com/humhub/humhub/issues/4828#issuecomment-773345719

yurabakhtin commented:
@luke- I have fixed the error of unknown method `getContainerPermissions()` in the commit https://github.com/humhub/humhub/pull/4833/commits/fe5838273e9b9e0dc58070c5eff953bdece29d92.

- I changed the method `getContainerPermissions()` of the parent/general class `ContentContainerModule` to public, because as I see all modules have the method as `public` and not `protected`, and after my first commit of this PR we have to keep this as `public`.
- Then as I understand all modules extended from `ContentContainerModule` must define permissions in the method `getContainerPermissions()` and not in the `getPermissions()`. It seems not all container modules were updated to use the method. It why I did the fix https://github.com/humhub/humhub/pull/4833/commits/fe5838273e9b9e0dc58070c5eff953bdece29d92#diff-3ebc21df74e6cf05d89de34d4e59c1492ae1182da2cc03fe232e6235775c46ceR33-R37

---

I think we should fix all `class Module extends ContentContainerModule` to use `getContainerPermissions()` instead of current:

```php
public function getPermissions($contentContainer = null)
{
    if (!$contentContainer) {
        return [
            new permissions\SomePermission(),
        ];
    }

    return [];
}
```

but some modules have the permissions only for Space or User container by condition like `if ($contentContainer instanceof Space) {`, so we cannot modify the modules easy, probably we should firstly modify core in order to display the permission depending on `BasePermission::defaultAllowedGroups`:

```php
protected $defaultAllowedGroups = [
    Space::USERGROUP_OWNER,
    Space::USERGROUP_ADMIN,
    User::USERGROUP_SELF,
];
```

As I understand the permission should be used only for defined `defaultAllowedGroups`, so it is strange why there is additional conditions like `if ($contentContainer instanceof Space)`.

luke- commented:
@yurabakhtin `defaultAllowedGroups`  indicates whether the permission is enabled by default for a user group. The permission can also be used for other user groups. What is the problem with the used condition `if ($contentContainer instanceof Space) {`?

The next step is to make sure that all modules really use the getContentContainerPermission.

yurabakhtin commented:
@luke- 

> What is the problem with the used condition `if ($contentContainer instanceof Space) {`?

I think the condition should be kept in the `getContainerPermissions()` if we will modify all `ContentContainerModule` to use it instead of `getPermissions()`.
I mean the condition `if (!$contentContainer) {` can be easy removed during the modification like it was done e.g. here https://github.com/humhub/humhub-modules-wiki/commit/6f2e9db1#diff-074f64cd19bc13bc527290951a58a65cb54f2f35a43ce5bf95e5dbb20b4d297dL73, but we should be more careful for the conditions like `if ($contentContainer instanceof Space)` because if we will remove them then the permission will become visible for other Container (User) that is wrong.

buddh4 commented on protected/humhub/modules/content/components/ContentContainerDefaultPermissionManager.php:
```diff
@@ -28,9 +28,21 @@ class ContentContainerDefaultPermissionManager extends PermissionManager
      */
     protected function getModulePermissions(\yii\base\Module $module)
     {
+        if ($module instanceof ContentContainerModule && method_exists($module, 'getContainerPermissions')) {
```
Think we can remove the `method_exists` here since isntanceof should make sure this method exists, hiding it in sub classes should not be allowed.

luke- replied:
https://github.com/humhub/humhub/commit/2059b55bc78fe127acfb0806b1e1c9df3a5e996e

buddh4 commented on protected/humhub/modules/content/components/ContentContainerDefaultPermissionManager.php:
```diff
@@ -28,9 +28,21 @@ class ContentContainerDefaultPermissionManager extends PermissionManager
+            if (!empty($containerPermissions)) {
+                // Don't try to find container permissions in the ContentContainerModule::getPermissions() below
+                // since they are already defined in more proper method ContentContainerModule::getContainerPermissions()
+                return $containerPermissions;
```
A ContentContainerModule can also export non container related permissions in `getPermissions()` think those won't be included in this logic?

buddh4 replied:
Also note that some container modules simply overwritten the `getPermissions()` instead of using the `getContainerPermissions()`

luke- replied:
Look good for me. Modules can either export permissions via `getContainerPermissions()` (preferred) or `getPremissions()`.

buddh4 replied:
Ok, this permission manager is solely used to determine default permissions in the admin settings.
